// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/atomic/badge/PillBox.js
import UIComponent from "../../../base";
import classNames from "classnames";
import { svg } from "../icon";
import { unique } from "../../../helpers/core/generators";
import { emitUsingProp } from "../../../helpers/props/props";
import i18n from "i18n";

const { content: { Clear } } = svg;
const NAMESPACE = "mpageui-PillBox";
const CLEAR = (new Clear().render());

/**
 * Renders the delete action for the PillBox.
 * @param {Function} el - A function for creating virtual elements.
 * @param {Function} onDelete - Function to call when the delete
 * action is clicked.
 * @param {display} display - To make it accessible
 * @returns {VDom} Rendered delete action.
 */
const renderDelete = (el, onDelete, display) => el(
    "a",
    {
        class: `${NAMESPACE}-clear mpageui-u-display-flex mpageui-u-align-items-center mpageui-u-pointer`,
        onclick: onDelete,
        role: "button",
        title: i18n().delete.replace("{0}", `${display}`),
        "aria-label": i18n().delete.replace("{0}", `${display}`)
    },
    CLEAR
);

/**
 * PillBox control class.
 * @class PillBox
 */
class PillBox extends UIComponent {
    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            display: "", // Required
            classNames: "",
            key: unique(NAMESPACE),
            requestDeleteEventName: null // Required
        };
    }

    /**
     * @inheritDoc
     */
    view(el, props, children) {
        return el(
            "div",
            {
                key: props.key,
                class: classNames(NAMESPACE, {
                    [ props.classNames ]: !!props.classNames
                })
            },
            [
                el(
                    "span",
                    {
                        class: `${NAMESPACE}-display mpageui-u-display-flex mpageui-u-align-items-center`
                    },
                    el(
                        "span",
                        {},
                        props.display
                    )
                ),
                renderDelete(
                    el,
                    emitUsingProp("requestDeleteEventName", this, this, { key: props.key }),
                    props.display
                )
            ]
        );
    }
}

export default PillBox;
