// MPageFusion/testing/src/pages/controls/composite/tag-list/TagList.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _seleniumWebdriver = require("selenium-webdriver");

var _UIComponent = _interopRequireDefault(require("../../../UIComponent.page"));

var _List = _interopRequireDefault(require("../../../helpers/List"));

var _Anchor = _interopRequireDefault(require("../../atomic/label/Anchor.page"));

var _Tags = _interopRequireDefault(require("./Tags.page"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var NAMESPACE = ".mpageui-TagList";
var SHOW_ALL_SELECTOR = NAMESPACE + "-all";
var SHOW_LESS_SELECTOR = NAMESPACE + "-preview";
var TAGS_SELECTOR = NAMESPACE + "-tags";
var SHOW_SELECTOR = NAMESPACE + "-show";
var ROOT_SELECTOR = NAMESPACE;
/**
 * Check that show all content is visible or not.
 * @param {TagListPage} tagListPage - The current instance of the tag list page object model
 * @returns {Promise<Boolean>} Boolean value stating whether the showAll is true
 * @private
 */

var isShowAllVisible = function isShowAllVisible(tagListPage) {
  return tagListPage.getRootElement().then(function (element) {
    return element.findElement(_seleniumWebdriver.By.css(SHOW_ALL_SELECTOR + " " + TAGS_SELECTOR));
  }).then(function (showAllElement) {
    return showAllElement.isDisplayed();
  })["catch"](function () {
    return false;
  });
};
/**
 * Check that show less content is visible or not.
 * @param {TagListPage} tagListPage - The current instance of the tag list page object model
 * @returns {Promise<Array>} Boolean value stating whether the showLess is true
 * @private
 */


var isShowLessVisible = function isShowLessVisible(tagListPage) {
  return tagListPage.getRootElement().then(function (element) {
    return element.findElement(_seleniumWebdriver.By.css(SHOW_LESS_SELECTOR));
  }).then(function (showLessElement) {
    return showLessElement.isDisplayed();
  })["catch"](function () {
    return false;
  });
};
/**
 * The TagList class
 * @class module:MPageFusion/testing.composite.entry.TagList
 *
 * @example
 * const tagList = new TagList(driver);
 * const list = await tagList.tags;
 * await list(0).get("display")) - give the tag label
 * await list(0).get("icon")) - gives the delete icon
 * const deleteIcon = await list(0).get("icon"));
 * await deleteIcon.click() - to delete the tag
 * const anchor = await tagList.showAllOrLessAnchor;
 * await anchor.click() - to click on showLess/showMore
 */


var TagListPage = /*#__PURE__*/function (_UIComponentPage) {
  _inheritsLoose(TagListPage, _UIComponentPage);

  function TagListPage() {
    return _UIComponentPage.apply(this, arguments) || this;
  }

  _createClass(TagListPage, [{
    key: "elementSelector",
    get:
    /**
     * @private
     */
    function get() {
      return ROOT_SELECTOR;
    }
    /**
     * Returns the list of displayed tags
     * @memberOf module:MPageFusion/testing.composite.entry.TagList
     * @returns {Promise<List<Tags>>} Returns the list of displayed tags or empty array.
     */

  }, {
    key: "tags",
    get: function get() {
      var _this = this;

      return Promise.all([isShowAllVisible(this), isShowLessVisible(this)]).then(function (_ref) {
        var _ref2 = _slicedToArray(_ref, 2),
            isShowAllElementVisible = _ref2[0],
            isShowLessElementVisible = _ref2[1];

        return !(isShowAllElementVisible || isShowLessElementVisible) ? Promise.resolve([]) : Promise.resolve((0, _List["default"])({
          list: _this,
          itemPage: _Tags["default"],
          selectorList: _this.selectorList.concat(isShowAllElementVisible ? SHOW_ALL_SELECTOR + " " + TAGS_SELECTOR : "" + SHOW_LESS_SELECTOR)
        }));
      });
    }
    /**
     * Returns the showAll/Less anchor element
     * @memberOf module:MPageFusion/testing.composite.entry.TagList
     * @returns {Promise<Anchor>} Returns showAll/showLess anchor element
     */

  }, {
    key: "showAllOrLessAnchor",
    get: function get() {
      var _this2 = this;

      return this.getRootElement().then(function (element) {
        return element.findElement(_seleniumWebdriver.By.css(SHOW_ALL_SELECTOR + " " + SHOW_SELECTOR));
      }).then(function (anchorElement) {
        return anchorElement.isDisplayed();
      }) // checks whether display property is not none.
      .then(function (isDisplayed) {
        if (!isDisplayed) {
          throw new Error("showAllOrLessAnchor will be displayed only in mobile and tablet mode");
        }

        return new _Anchor["default"](_this2.driver, {
          selectorList: _this2.selectorList.concat(SHOW_ALL_SELECTOR + " " + SHOW_SELECTOR),
          // TagListControl does not use the atomic/label/Anchor control. It uses its own "a" element.
          useExactSelector: true
        });
      });
    }
  }]);

  return TagListPage;
}(_UIComponent["default"]);

exports["default"] = TagListPage;
