// MPageFusion/src/main/js/cerner/mpagedev/mpagefusion/controls/composite/viewpoint-tab-list/ViewpointAddTab.js
import UIComponent from "../../../base";
import { Tab } from "../../atomic/tabs";
import { svg } from "../../atomic/icon";
import classNames from "classnames";
import i18n from "i18n";

const { content: { Add } } = svg;
const { addView: addViewI18n } = i18n();

const namespace = "mpageui-ViewpointAddTab";

const ViewpointAddTabEvents = {
    CLICK: "ViewpointAddTab::click"
};

/**
 * The PlusAddLink class. This is is a private class that is injected into
 * a Tab via the ViewpointPlusAddTab class. Not intended for external use.
 * @class PlusAddLink
 */
class PlusAddLink extends UIComponent {
    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            display: ""
        };
    }
    /**
     * @inheritDoc
     */
    createChildren() {
        return [
            new Add({
                classNames: `${namespace}-addIcon`
            })
        ];
    }

    /**
     * @inheritDoc
     */
    view(el, props, children) {
        return el(
            "a",
            {
                class: `${namespace}-link`,
                role: "button",
                title: addViewI18n,
                "aria-label": addViewI18n
            },
            [
                children [ 0 ].render(),
                el(
                    "span",
                    {
                        class: classNames(`${namespace}-display`, "mpageui-u-hide-for-md")
                    },
                    props.display
                )
            ]
        );
    }
}

/**
 * The ViewpointAddTab class.
 * @class ViewpointAddTab
 */
class ViewpointAddTab extends UIComponent {
    /**
     * @inheritDoc
     */
    initialProps() {
        return {
            display: "",
            id: ""
        };
    }

    /**
     * @inheritDoc
     */
    createChildren() {
        return [
            {
                tab: new Tab(
                    {
                        id: this.getProp("id"),
                        defaultClassNames: "mpageui-ViewpointAddTab",
                        draggable: false,
                        clickEventName: ViewpointAddTabEvents.CLICK
                    },
                    [
                        new PlusAddLink({
                            display: this.getProp("display")
                        })
                    ]
                )
            }
        ];
    }

    /**
     * @inheritDoc
     */
    view(el, props, children) {
        return this.getChild("tab").render();
    }
}

ViewpointAddTab.EVENTS = ViewpointAddTabEvents;

export default ViewpointAddTab;
