// MPageFusion/testing/src/pages/controls/composite/form/controls/FormCheckboxGroup.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _FormControl = _interopRequireDefault(require("../FormControl.page"));

var _CheckboxGroup = _interopRequireDefault(require("../../../composite/select-group/CheckboxGroup.page"));

var _helpers = require("../helpers");

var _ClinicalBadge = _interopRequireDefault(require("../../../atomic/badge/ClinicalBadge.page"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var ROOT_SELECTOR = ".mpageui-FormCheckboxGroup";
/**
 * The FormCheckboxGroupPage class
 * @class module:MPageFusion/testing.composite.form.FormCheckboxGroupPage
 *
 * @example
 * const formCheckboxGroup = new FormCheckboxGroupPage(driver);
 * formCheckboxGroup.get("display")
 *     .then((display) => {
 *         assert(display === "Sample Label");
 *     });
 */

var FormCheckboxGroupPage = /*#__PURE__*/function (_FormControlPage) {
  _inheritsLoose(FormCheckboxGroupPage, _FormControlPage);

  function FormCheckboxGroupPage(driver, options) {
    var _this;

    _this = _FormControlPage.call(this, driver, options) || this;
    _this._checkboxGroup = new _CheckboxGroup["default"](driver, {
      selectorList: _this.selectorList.concat(_helpers.CLASSES.RIGHT_COLUMN)
    });
    return _this;
  }

  var _proto = FormCheckboxGroupPage.prototype;

  /**
   * Toggles the checkbox at the specified index
   * @memberOf module:MPageFusion/testing.composite.form.FormCheckboxGroupPage
   * @param {Number} index - the index of the checkbox to toggle
   * @returns {Promise} resolves when the action completes
   */
  _proto.toggle = function toggle(index) {
    return this._checkboxGroup.checkbox(index).click();
  };

  _createClass(FormCheckboxGroupPage, [{
    key: "elementSelector",
    get: function get() {
      return ROOT_SELECTOR;
    }
    /**
     * Gets the current values in the checkbox group
     * @memberOf module:MPageFusion/testing.composite.form.FormCheckboxGroupPage
     * @returns {Promise<Array>} resolves with an array of objects { display, isSelected }
     */

  }, {
    key: "value",
    get: function get() {
      return this._checkboxGroup.checkbox.all().then(function (checkboxGroup) {
        return Promise.all(checkboxGroup.map(function (checkbox) {
          return Promise.all([checkbox.get("display"), checkbox.get("isSelected")]).then(function (_ref) {
            var _ref2 = _slicedToArray(_ref, 2),
                display = _ref2[0],
                isSelected = _ref2[1];

            return {
              display: display,
              isSelected: isSelected
            };
          });
        }));
      });
    }
    /**
     * returns the base CheckboxGroup page
     * @memberOf module:MPageFusion/testing.composite.form.FormCheckboxGroupPage
     * @returns {CheckboxGroupPage} checkboxGroupPage
     */

  }, {
    key: "checkboxGroup",
    get: function get() {
      return this._checkboxGroup;
    }
  }]);

  return FormCheckboxGroupPage;
}(_FormControl["default"]);

exports["default"] = FormCheckboxGroupPage;
FormCheckboxGroupPage.CRITICALITY_TYPES = _ClinicalBadge["default"].TYPES;
