// MPageFusion/testing/src/pages/controls/composite/list/ListButton.page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _UIComponent = _interopRequireDefault(require("../../../UIComponent.page"));

var _helpers = require("../../atomic/icon/svg/helpers");

var _seleniumWebdriver = require("selenium-webdriver");

var _helpers2 = require("./helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var ROOT_SELECTOR = ".mpageui-ListButton";
var BUTTON_ENABLED = "mpageui-ListButton--enabled";
var BUTTON_TYPE_SUBLIST = "mpageui-Button--sublist";
var BUTTON_TYPE_SELECTABLE = "mpageui-Button--listselect";
var BUTTON_TYPE_SELECT = "mpageui-listselectgroup-item";
/**
 * The ListButtonPage class
 * @class ListButtonPage
 * @private
 */

var ListButtonPage = /*#__PURE__*/function (_UIComponentPage) {
  _inheritsLoose(ListButtonPage, _UIComponentPage);

  function ListButtonPage() {
    return _UIComponentPage.apply(this, arguments) || this;
  }

  var _proto = ListButtonPage.prototype;

  /**
   * Clicks the button
   * @returns {Promise} Resolves when the action is complete
   */
  _proto.click = function click() {
    return this.getRootElement().then(function (element) {
      return element.click();
    });
  };

  _createClass(ListButtonPage, [{
    key: "elementSelector",
    get: function get() {
      return ROOT_SELECTOR;
    }
    /**
     * The display on the button
     * @returns {Promise<String>} display
     */

  }, {
    key: "display",
    get: function get() {
      return this.getRootElement().then(function (element) {
        return element.getText();
      });
    }
    /**
     * The icon on the button
     * @returns {Promise<String>} icon
     */

  }, {
    key: "icon",
    get: function get() {
      return this.getRootElement().then(function (element) {
        return element.findElement(_seleniumWebdriver.By.css(".mpageui-SvgIcon"));
      }).then(function (element) {
        return element.getAttribute("class");
      }).then(function (classNames) {
        var iconClass = classNames.split(/\s+/).find(function (className) {
          return className.includes("mpageui-SvgIcon-");
        });
        return _helpers.classTypeMap[iconClass];
      })["catch"](function () {
        return "";
      });
    }
    /**
     * Whether the button is enabled or not
     * @returns {Promise<Boolean>} isEnabled
     */

  }, {
    key: "isEnabled",
    get: function get() {
      return this.getRootElement().then(function (element) {
        return element.getAttribute("class");
      }).then(function (classNames) {
        return classNames.includes(BUTTON_ENABLED);
      });
    }
    /**
     * Whether the button is selected or not
     * @returns {Promise<Boolean>} isSelected
     */

  }, {
    key: "isSelected",
    get: function get() {
      return this.getRootElement().then(function (element) {
        return element.getAttribute("aria-selected");
      }).then(function (attribute) {
        return Boolean(attribute);
      });
    }
    /**
     * The type of the button
     * @returns {Promise<String>} type
     */

  }, {
    key: "type",
    get: function get() {
      return this.getRootElement().then(function (element) {
        return element.getAttribute("class");
      }).then(function (classNames) {
        if (classNames.includes(BUTTON_TYPE_SELECTABLE)) {
          if (classNames.includes(BUTTON_TYPE_SELECT)) {
            return _helpers2.BUTTON_TYPES.SELECT;
          }

          return _helpers2.BUTTON_TYPES.TOGGLE;
        }

        if (classNames.includes(BUTTON_TYPE_SUBLIST)) {
          return _helpers2.BUTTON_TYPES.SUBLIST;
        }

        return _helpers2.BUTTON_TYPES.LIST;
      });
    }
    /**
     * The object representing the state of the button
     * @returns {Promise<Object>} value
     */

  }, {
    key: "value",
    get: function get() {
      return Promise.all([this.get("display"), this.get("type"), this.get("icon"), this.get("isEnabled"), this.get("isSelected")]).then(function (_ref) {
        var _ref2 = _slicedToArray(_ref, 5),
            display = _ref2[0],
            type = _ref2[1],
            icon = _ref2[2],
            isEnabled = _ref2[3],
            isSelected = _ref2[4];

        return {
          display: display,
          type: type,
          icon: icon,
          isEnabled: isEnabled,
          isSelected: isSelected
        };
      });
    }
  }]);

  return ListButtonPage;
}(_UIComponent["default"]);

exports["default"] = ListButtonPage;
